hypervisor/dm/vioapic.c

```diff
@@ -85,91 +85,69 @@ vioapic_send_intr(struct acrn_vioapic *vioapic, uint32_t pin)
 
 /**
  * @pre pin < vioapic_pincount(vm)
+ *
+ * here the assumption is that IRQ only fires at raising edge for edge
+ * triggered interrupt, and high level for level triggered interrupt.
+ * actually, the virtual interrupt asserting for vioapic does not consider
+ * the polarity of the RTE setting.
  */
```

Comment:
Then please check the guest write of RTE to assert or fail on guest change of polarity.

Reply:
will update

Reply:
@dongyaozu , there is one problem here, for ptdev, we need update polarity configuration from vioapic to native ioapic, so we cannot assert guest writing RTE for polarity setting.  
Per my view, we can emulate virtual pin state in vioapic fully, but it may bring other complex, for example:
if the assumption is like
- vioapic_assert_irq -> pin_state to high level
- vioapic_deassert_irq -> pin_state to low level
then how should we trigger a low level active interrupt? We should keep the pin_state as high at the beginning, and the emulated device should call vioapic_deassert_irq to trigger interrupt, and then call vioapic_assert_irq to stop triggering.

Reply:
I am suggesting to support one polarity only -- this is what your patch assumes. In this case, if the guest change polarity, we should assert. I don't see any issue.

If we fully emulate the polarity, and propagate the vioapic polarity to physical IOAPIC. At the time of PIRQ, we propagate the physical IOAPIC level to vioapic level.  That works fine too.

Reply:
For PTDEV INTx interrupt, we must keep polarity configuration from vioapic as it must be set to native ioapic accordingly, otherwise, this device may fail to work. So we cannot support one polarity only.

And yes, we can fully emulate the polarity, but it bring complicated change for both ptdev and emulated device irq assert/deassert. My question is that, why do we must take care polarity, when there is interrupt come, then we just assert it and trigger corresponding virtual interrupt, it should be enough.